contrib/Shopify/sarama/sarama.go

```diff
@@ -178,56 +180,62 @@ func WrapAsyncProducer(saramaConfig *sarama.Config, p sarama.AsyncProducer, opts
 	}
 	if saramaConfig == nil {
 		saramaConfig = sarama.NewConfig()
+		saramaConfig.Version = sarama.V0_11_0_0
+	}
+
+	headersSupported := true
+	if !saramaConfig.Version.IsAtLeast(sarama.V0_11_0_0) {
+		headersSupported = false
+		log.Println("Tracing Sarama async producer requires at least sarama.V0_11_0_0 version")
```

Comment:
using `log` like this in libraries is very dirty
what do you think of changing `WrapAsyncProducer` to also return and `error` ?
generally, one of the reasons the "functional options" pattern became popular is that it can be used for validation as well

reference: https://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis

TBH I do not understand why even OTel lib is not having the `error` in the API even if it would be always `nil`. it just reduces potential risk of introducing a breaking change

Reply:
Generally instrumentations don't raise/return errors and silently become noop implementations when they can't work properly. Go is a bit different as all instrumentations have to be applied manually but still so far they don't halt program execution and at most log warnings. I don't want to introduce a change like that now. This shouldn't be triggered in reality TBH as 0.10 is very very old.

Reply:
Also, in this case the instrumentation still works and generates spans, it just does so a bit less accurately.

Reply:
> Generally instrumentations don't raise/return errors and silently become noop implementations when they can't work properly

I do not think it is good for incorrect configuration. I think the API consumer should be able to know that something is wrong.

Take notice https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/error-handling.md#basic-error-handling-principles:

> The API or SDK may fail fast and cause the application to fail on initialization,

And doing a side effect like `log.Println("Tracing Sarama async producer requires at least sarama.V0_11_0_0 version")` is very bad as it is very tricky for the user to disable it (or change) if needed. If we want to log we should at least do make some adapter that the user can use to inject its own logger.

However, for https://github.com/signalfx/signalfx-go-tracing we can leave it as it is. I see that `log` is used in some places so this approach is consistent. Feel free to resolve 👍

Reply:
> Take notice https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/error-handling.md#basic-error-handling-principles:

1. This is not really Otel :)
2. Even if this was, this isn't part of either the SDK or the API since it's an instrumentation and generally in Otel instrumentations never result in crashes. It is more or less acceptable to do it during initialization but still in this case I don't think we should completely lock out users still on older version. We have a fallback in place along with a warning log. I don't see how refusing to work instead of providing slightly degraded experience (incorrect span end time) is better TBH.

Reply:
> If we want to log we should at least do make some adapter that the user can use to inject its own logger.

Yea but most I can do about it today is create an issue :)

Reply:
Resolving 😉 

> Yea but most I can do about it today is create an issue 

Maybe only if some client would complain or require it 😉